### PR TITLE
`azurerm_dev_test_virtual_network`: add support for `shared_public_ip_address_configuration`

### PR DESCRIPTION
**Pull request by jiaweitao001**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [x] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
=== RUN   TestAccDevTestVirtualNetwork_subnet
=== PAUSE TestAccDevTestVirtualNetwork_subnet
=== CONT  TestAccDevTestVirtualNetwork_subnet
--- PASS: TestAccDevTestVirtualNetwork_subnet (477.24s)
PASS
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

`azurerm_dev_test_virtual_network`: add support for `shared_public_ip_address_configuration`

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #0000

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go [hunk 1]
```diff
@@ -96,6 +97,37 @@ func resourceArmDevTestVirtualNetwork() *pluginsdk.Resource {
 							Default:      string(virtualnetworks.UsagePermissionTypeAllow),
 							ValidateFunc: validate.DevTestVirtualNetworkUsagePermissionType(),
 						},
+
+						"shared_public_ip_address_configuration": {
```
Comment (stephybun) [inline internal/services/devtestlabs/dev_test_virtual_network_resource.go:101]: `configuration` is redundant here, this should be renamed to 
```suggestion
						"shared_public_ip_address": {
```

File: internal/services/devtestlabs/dev_test_virtual_network_resource_test.go [hunk 2]
```diff
@@ -203,6 +203,12 @@ resource "azurerm_dev_test_virtual_network" "test" {
   subnet {
     use_public_ip_address           = "Deny"
     use_in_virtual_machine_creation = "Allow"
+    shared_public_ip_address_configuration {
+      allowed_ports {
+        backend_port       = 22
+        transport_protocol = "Tcp"
+      }
```
Comment (stephybun) [inline internal/services/devtestlabs/dev_test_virtual_network_resource_test.go:210]: Multiple can be specified, can we please add another one to the test as well as test the update of these?

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @jiaweitao001, would you mind checking the comments left in-line?

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go (hunk 1 above)
Comment (jiaweitao001) [reply internal/services/devtestlabs/dev_test_virtual_network_resource.go:101]: Will do.

File: internal/services/devtestlabs/dev_test_virtual_network_resource_test.go (hunk 2 above)
Comment (jiaweitao001) [reply internal/services/devtestlabs/dev_test_virtual_network_resource_test.go:210]: Will add.

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go [hunk 3]
```diff
@@ -305,18 +336,53 @@ func expandDevTestVirtualNetworkSubnets(input []interface{}, subscriptionId, res
 
 	for _, val := range input {
 		v := val.(map[string]interface{})
+		sharedPublicIpAddressConfiguration := expandDevTestVirtualNetworkSubnetIpAddressConfiguration(v["shared_public_ip_address"].([]interface{}))
```
Comment (stephybun) [inline internal/services/devtestlabs/dev_test_virtual_network_resource.go:339]: There isn't a need to declare this as a separate variable.
```suggestion
```

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go [hunk 4]
```diff
@@ -305,18 +336,53 @@ func expandDevTestVirtualNetworkSubnets(input []interface{}, subscriptionId, res
...
 		subnet := virtualnetworks.SubnetOverride{
-			ResourceId:                   pointer.To(subnetId.ID()),
-			LabSubnetName:                pointer.To(name),
-			UsePublicIPAddressPermission: pointer.To(virtualnetworks.UsagePermissionType(v["use_public_ip_address"].(string))),
-			UseInVMCreationPermission:    pointer.To(virtualnetworks.UsagePermissionType(v["use_in_virtual_machine_creation"].(string))),
+			ResourceId:                         pointer.To(subnetId.ID()),
+			LabSubnetName:                      pointer.To(name),
+			UsePublicIPAddressPermission:       pointer.To(virtualnetworks.UsagePermissionType(v["use_public_ip_address"].(string))),
+			UseInVMCreationPermission:          pointer.To(virtualnetworks.UsagePermissionType(v["use_in_virtual_machine_creation"].(string))),
+			SharedPublicIPAddressConfiguration: sharedPublicIpAddressConfiguration,
```
Comment (stephybun) [inline internal/services/devtestlabs/dev_test_virtual_network_resource.go:346]:
```suggestion
			SharedPublicIPAddressConfiguration: expandDevTestVirtualNetworkSubnetIpAddressConfiguration(v["shared_public_ip_address"].([]interface{})),
```

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go [hunk 5]
```diff
@@ -305,18 +336,53 @@ func expandDevTestVirtualNetworkSubnets(input []interface{}, subscriptionId, res
...
 }
 
+func expandDevTestVirtualNetworkSubnetIpAddressConfiguration(input []interface{}) *virtualnetworks.SubnetSharedPublicIPAddressConfiguration {
+	if len(input) == 0 {
+		return nil
+	}
+
+	v := input[0].(map[string]interface{})
+	allowedPortsRaw := v["allowed_ports"].([]interface{})
+	allowedPorts := expandDevTestVirtualNetworkSubnetAllowedPorts(allowedPortsRaw)
```
Comment (stephybun) [inline internal/services/devtestlabs/dev_test_virtual_network_resource.go:361]: Same here 
```suggestion
	allowedPorts := expandDevTestVirtualNetworkSubnetAllowedPorts(v["allowed_ports"].([]interface{}))
```

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go [hunk 6]
```diff
@@ -305,18 +336,53 @@ func expandDevTestVirtualNetworkSubnets(input []interface{}, subscriptionId, res
...
+
+	for _, val := range input {
+		v := val.(map[string]interface{})
+		backendPort := int64(v["backend_port"].(int))
+		transportProtocol := virtualnetworks.TransportProtocol(v["transport_protocol"].(string))
+
+		allowedPort := virtualnetworks.Port{
+			BackendPort:       pointer.To(backendPort),
+			TransportProtocol: pointer.To(transportProtocol),
+		}
```
Comment (stephybun) [inline internal/services/devtestlabs/dev_test_virtual_network_resource.go:379]: Same here 
```suggestion
		allowedPort := virtualnetworks.Port{
			BackendPort:       pointer.To(int64(v["backend_port"].(int))),
			TransportProtocol: pointer.To(virtualnetworks.TransportProtocol(v["transport_protocol"].(string))),
		}
```

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go [hunk 7]
```diff
@@ -330,9 +396,45 @@ func flattenDevTestVirtualNetworkSubnets(input *[]virtualnetworks.SubnetOverride
...
+	}
+
+	for _, v := range *input {
+		output := make(map[string]interface{})
+		if v.BackendPort != nil {
+			output["backend_port"] = v.BackendPort
+		}
+		if v.TransportProtocol != nil {
+			output["transport_protocol"] = v.TransportProtocol
+		}
```
Comment (stephybun) [inline internal/services/devtestlabs/dev_test_virtual_network_resource.go:435]: These could be condensed using `pointer.From`
```suggestion
		output["backend_port"] = pointer.From(v.BackendPort)
		output["transport_protocol"] = pointer.From(v.TransportProtocol)
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @jiaweitao001, there are a few additional comments. Once they're resolved this should be good to go.

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go (hunk 3 above)
Comment (jiaweitao001) [reply internal/services/devtestlabs/dev_test_virtual_network_resource.go:339]: Will change.

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go (hunk 4 above)
Comment (jiaweitao001) [reply internal/services/devtestlabs/dev_test_virtual_network_resource.go:346]: Will change.

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go (hunk 5 above)
Comment (jiaweitao001) [reply internal/services/devtestlabs/dev_test_virtual_network_resource.go:361]: Will change.

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go (hunk 6 above)
Comment (jiaweitao001) [reply internal/services/devtestlabs/dev_test_virtual_network_resource.go:379]: Will change.

File: internal/services/devtestlabs/dev_test_virtual_network_resource.go (hunk 7 above)
Comment (jiaweitao001) [reply internal/services/devtestlabs/dev_test_virtual_network_resource.go:435]: Will change.

**Review by stephybun — APPROVED** [review]
Thanks @jiaweitao001 LGTM 👍

**Pull request merged**